contracts/interchain-token-service/src/contract/execute.rs

```diff
@@ -186,16 +209,179 @@ pub fn set_chain_config(
     max_uint: nonempty::Uint256,
     max_target_decimals: u8,
 ) -> Result<Response, Error> {
-    match state::may_load_chain_config(deps.storage, &chain)
-        .change_context_lazy(|| Error::LoadChainConfig(chain.clone()))?
-    {
+    match state::may_load_chain_config(deps.storage, &chain).change_context(Error::State)? {
         Some(_) => bail!(Error::ChainConfigAlreadySet(chain)),
         None => state::save_chain_config(deps.storage, &chain, max_uint, max_target_decimals)
-            .change_context_lazy(|| Error::SaveChainConfig(chain))?
+            .change_context(Error::State)?
             .then(|_| Ok(Response::new())),
     }
 }
 
+/// Applies various checks on the message.
+fn apply_checks(
+    storage: &mut dyn Storage,
+    message: &Message,
+    message_direction: MessageDirection,
+) -> Result<(), Error> {
+    let token_chain_info = state::may_load_token_chain_info(
+        storage,
+        message_direction.clone().into(),
+        message.token_id(),
+    )
+    .change_context(Error::State)?;
+
+    // Validation checks
+    chain_frozen_check(storage, &message_direction.clone().into())?;
+    token_redeployment_check(message, &message_direction, &token_chain_info)?;
+    token_deployment_origin_chain_check(storage, message, &message_direction, &token_chain_info)?;
+
+    // Transformations
+    let token_chain_info = token_deployment_check(message, &message_direction, token_chain_info)?
+        .then(|token_chain_info| {
+        token_supply_check(message, &message_direction, token_chain_info)
+    })?;
+
+    state::save_token_chain_info(
+        storage,
+        message_direction.into(),
+        message.token_id(),
+        &token_chain_info,
+    )
+    .change_context(Error::State)
+}
+
+/// Ensures that the token is being redeployed from the same origin chain.
+/// If it's being deployed for the first time, the from chain is saved as the origin chain in the token config.
+fn token_deployment_origin_chain_check(
```

Comment:
```suggestion
fn check_token_deployment_origin_chain(
```
Actually, `check` isn't even a good name here as this function writes to the storage but `check` suggests that it does not.

Reply:
Yeah, they are a mixture of checks and transformations. I'm thinking whether we need any consistency to identify all such functions (in which case we can name each in a more apt way), or we use some other word like `handler` instead of `check`